Add unnamed scripts as an option to create dialogs.

tle-oops opened the pull request:
When `class_name` isn't an option for gdscript classes for your project for any reason. This change list unnamed scripts within the base type's class hierarchy to create dialogs.

Closes: https://github.com/godotengine/godot-proposals/issues/8102
Here's an example of what the create-dialogs will look like listing unnamed scripts. 
First screenshot includes script that extend an engine class that extends one of the Resource class.
Second screenshot includes scripts that directly inherit the Resource or another script that is part of that hierarchy.
<img width="626" height="204" alt="image" src="https://github.com/user-attachments/assets/12d95d76-da91-4544-b34c-275aaff62cfc" />
<img width="810" height="284" alt="image" src="https://github.com/user-attachments/assets/7c5a92d7-1541-439a-a679-77f7b80524a4" />

tle-oops committed:
Added unnamed scripts as an option to create dialogs.

tle-oops commented:
linked issue: https://github.com/godotengine/godot/pull/109784

tle-oops commented:
Wasn't aware drafts also send out notifications.

AThousandShips commented:
It doesn't generally, but regardless please don't create duplicate PRs

ogapo commented:
@AThousandShips sorry for the buzz. @tle-oops is on my team and I asked him to open a draft PR so I could review this before it got sent over yall's way in an effort to reduce the work required on your end. Clearly that didn't work out as intended, appologies. We thought a draft wouldn't trigger any notifications on your side. What's the recommended workflow in this case?

AThousandShips commented:
Any PR opened here has to be processed, so opening a PR here should be done when wanting to get *our* attention on it, or getting feedback from us, but at minimum it should be communicated that the PR is just for feedback

But generally please don't use this repo for internal review or feedback

tle-oops committed:
Fix style check failure.

tle-oops committed:
added empty name check for type list.

tle-oops committed:
removed unused var

tle-oops committed:
rename script var to fix compilation error

ogapo commented on editor/gui/create_dialog.cpp:
```diff
@@ -395,12 +426,19 @@ void CreateDialog::_configure_search_option_item(TreeItem *r_item, const StringN
+		meta.append(is_unnamed_script);
+	}
+
+	r_item->set_metadata(0, meta);
```
We should ensure that there's always a consistent number of metadata entries by setting them all right before adding. Right now the 3rd entry isn't added at all unless is_unnamed_script is true which might trigger the assert on line 675.
```cpp
Array meta;
meta.append(is_custom_type);
meta.append(type_name);
meta.append(is_unnamed_script);
r_item->set_metadata(0, meta);
```

ogapo commented on editor/gui/create_dialog.cpp:
```diff
@@ -131,6 +135,26 @@ void CreateDialog::_fill_type_list() {
+	}
+
+	for (int i = 0; i < fsd->get_file_count(); i++) {
+		if (fsd->get_file_script_class_extends(i) == base_type && fsd->get_file_script_class_name(i).is_empty()) {
```
I think there's 3 case we should handle:
1. The script could extend from the base_type directly -- handled here
2. The script could extend another named class that itself extends base_type -- need to use `ClassDB::is_parent_class` to handle this
3. The script could extend another script class (and turtles all the way down) somewhere at the end extending a named type -- should be able to check if the extends starts with `res://` I think?

tle-oops committed:
Updated to search through class doc.
Added param to create_dialog::popup_create() to allow optional inclusion of unnamed scripts.
Currently only Inspector dock (Resource) includes by default.

tle-oops commented:
Here's what the resource picker will look like when including unnamed scripts that's part of the base_type inheritance tree.

<img width="626" height="204" alt="Screenshot 2025-08-21 at 12 56 23 PM" src="https://github.com/user-attachments/assets/58ee2533-fb41-40f1-994b-8b30cd1cb670" />
<img width="810" height="284" alt="Screenshot 2025-08-21 at 12 56 42 PM" src="https://github.com/user-attachments/assets/7c96f0d0-758f-458a-8645-defc5d5bfdbb" />

tle-oops committed:
Merge branch 'master' into tle/resource-picker
